File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        dict = {'(':')', '{':'}', '[':']'}
        stack = []
        for i in s:
            if i in dict:
                stack.append(i)

            # Means if the stack is empty or the closing bracket doesn't match the 
            # last opening bracket then return False
            elif not stack or dict[stack.pop()] != i: 
                return False
        return len(stack) == 0

File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}]
 */
var isValid = function (s) {
    if (s.length % 2 !== 0) return false;

    const map = new Map([
        [")", "("],
        ["]", "["],
        ["}", "{"],
    ]);

    let stack = [];
    let open = new Set(["(", "[", "{"]);

    for (const char of s) {
        if (open.has(char)) stack.push(char);
        else if (stack.pop() !== map.get(char)) return false;
    }

    return stack.length === 0;
};